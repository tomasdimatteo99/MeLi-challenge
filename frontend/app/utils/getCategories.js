const axios = require('axios');
const { notFound } = require('next/navigation');

//Mover al backend.
const getCategories = async( id ) =>{
  try{
      const response  = await axios.get(`https://api.mercadolibre.com/categories/${id}`);
      const categories = response.data.path_from_root.map( catVal => catVal.name);
      return {categories: categories};
  }catch(err){
      notFound()
  }
}

module.exports = { getCategories };
