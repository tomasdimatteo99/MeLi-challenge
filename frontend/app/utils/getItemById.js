const axios = require('axios');
const { notFound } = require('next/navigation');

const getItemById = async( id ) =>{
  try{
      const response  = await axios.get(`http://localhost:8080/api/items/${id}`);
      return response.data;
  }catch(err){
      notFound()
  }
}

module.exports = { getItemById };