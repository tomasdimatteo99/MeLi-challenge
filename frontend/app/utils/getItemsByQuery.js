const axios = require('axios');
const { notFound } = require('next/navigation');

const getItemsByQuery = async(searchParams = "iphone") =>{
    try{
        const response  = await axios.get(`http://localhost:8080/api/items?q=${searchParams}`);
        return response.data;
    }catch(err){
        notFound()
    }
}

module.exports = { getItemsByQuery };