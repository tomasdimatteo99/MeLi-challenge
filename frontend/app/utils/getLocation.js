const axios = require('axios');
const { notFound } = require('next/navigation');

//Mover al backend.
const getLocation = async( id ) =>{
  try{
      const response  = await axios.get(`https://api.mercadolibre.com/items/${id}`);
      const location = response.data.seller_address.state.name;
      return {location: location};
  }catch(err){
      notFound()
  }
}

module.exports = { getLocation };
