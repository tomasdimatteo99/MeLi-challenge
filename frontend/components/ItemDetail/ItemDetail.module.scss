@use "../../styles/sass/variable/mediaQueries" as mediaQuery;
@import '../../styles/sass/globals.scss';

.divContainer{
    .itemDetailContainer{
        @include container();
        @include flexRow();
        background-color: white;
        border-radius: 4px;
        margin-bottom: 32px;
    
        .leftColumn{
            @include container($width:60%);
            @include flexColumn($alingItems:flex-start);
            @include setPadding($padding: 32px);
        }
    
        .rightColumn{
            @include container($width:40%);
            @include flexColumn($alingItems:flex-start);
            @include setPadding($padding: 32px);
            margin: 0 0 auto;
        }
    
<<<<<<< HEAD
        @media #{mediaQuery.$md} {
            @include container($height:100%);
=======
        @media #{mediaQuery.$xl} {
            @include container($width:100vw ,$height:100%);
    }   
}
