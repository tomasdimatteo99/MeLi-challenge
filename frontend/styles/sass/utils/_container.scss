@use "../variable/mediaQueries" as mediaQuery;

//Setters
@mixin setWidth($width){
    max-width: $width;
    min-width: $width;
    width: $width;
}

@mixin setHeight($height){
    max-height: $height;
    min-height: $height;
    height: $height;
}

@mixin setDimension(
    $width,
    $height
){
    @include setWidth($width);
    @include setHeight($height);
}

@mixin setMargin(
    $margin:0,
    $margin-l:$margin,
    $margin-r:$margin,
    $margin-b:$margin,
    $margin-t:$margin
){
    margin-left: $margin-l;
    margin-right: $margin-r;
    margin-bottom: $margin-b;
    margin-top: $margin-t;
}

@mixin setMargin-x($margin-x){
    margin-left: $margin-x;
    margin-right: $margin-x;
}

@mixin setMargin-y($margin-y){
    margin-top: $margin-y;
    margin-bottom: $margin-y;
}

@mixin setPadding(
    $padding:0,
    $padding-l:$padding,
    $padding-r:$padding,
    $padding-b:$padding,
    $padding-t:$padding
){
    padding-left: $padding-l;
    padding-right: $padding-r;
    padding-bottom: $padding-b;
    padding-top: $padding-t;
}

@mixin setPadding-x($padding-x){
    padding-left: $padding-x;
    padding-right: $padding-x;
}

@mixin setPadding-y($padding-y){
    padding-top: $padding-y;
    padding-bottom: $padding-y;
}

//Modeling
@mixin container-fluid(
    $padding-x: 14px,
    $margin-x: auto,
    $width: 100vw,
    $height: auto
){
    @include setMargin-x($margin-x);
    @include setPadding-x($margin-x);
    @include setWidth($width);
    @include setHeight($height);

    //Small
    @media #{mediaQuery.$sm} {
        @include setMargin-x($margin-x);
        @include setPadding-x($margin-x);
        @include setWidth($sm-px);
    }
    //Medium
    @media #{mediaQuery.$md} {
        @include setMargin-x($margin-x);
        @include setPadding-x($margin-x);
        @include setWidth($md-px);
    }
    //Large
    @media #{mediaQuery.$lg} {
        @include setMargin-x($margin-x);
        @include setPadding-x($margin-x);
        @include setWidth($lg-px);
    }
    //Extra - Large
    @media #{mediaQuery.$xl} {
        @include setMargin-x($margin-x);
        @include setPadding-x($margin-x);
        @include setWidth($xl-px);
    }
    //Extra extra - Large
    @media #{mediaQuery.$xxl} {
        @include setMargin-x($margin-x);
        @include setPadding-x($margin-x);
        @include setWidth($xxl-px);
    }
}

@mixin container(
    $padding-x: 14px,
    $margin-x: auto,
    $width: 80vw,
    $height: auto
){
    @include container-fluid(
        $padding-x,
        $margin-x,
        $width,
        $height
    );
}

@mixin simpleContainer(
    $padding-x: 14px,
    $margin-x: auto,
    $width: 100vw,
    $height: auto
){
    @include setMargin-x($margin-x);
    @include setPadding-x($margin-x);
    @include setWidth($width);
    @include setHeight($height);
}