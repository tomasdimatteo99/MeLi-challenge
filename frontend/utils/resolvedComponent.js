const resolvedComponent = async (Component, props) => {
    const ComponentResolved = await Component(props);
    return () => ComponentResolved;
}

module.exports = { resolvedComponent };