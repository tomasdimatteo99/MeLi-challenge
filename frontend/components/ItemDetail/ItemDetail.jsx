import BuyButton from "./BuyButton/BuyButton";
import ItemDetailData from "./ItemDetailData/ItemDetailData";
import ItemDetailDescription from "./ItemDetailDescription/ItemDetailDescription";
import ItemDetailImage from "./ItemDetailImage/ItemDetailImage";
import ItemCategory from "../ItemCategory/ItemCategory";
import style from "./ItemDetail.module.scss"
import { getCategories } from "../../app/utils/getCategories";

export default async function ItemDetail( props ) {
  const id = props.item.category_id;
  const categories = await getCategories( id );
  return (
    <div className={style.divContainer}>
      <ItemCategory itemsCategories={categories.categories}/>
      <section className={style.itemDetailContainer}>
        <div className={style.leftColumn}>
          <ItemDetailImage
            picture={props.item.picture}
            title={props.item.title}
          />
          <ItemDetailDescription
            description={props.description}
          />
        </div>
        <div className={style.rightColumn}>
          <ItemDetailData
            condition={props.item.condition}
            soldQuantity={props.item.sold_quantity}
            title={props.item.title}
            currency={props.item.price.currency}
            amount={props.item.price.amount}
          />
          <BuyButton/>
        </div>
      </section>
    </div>
  );
}