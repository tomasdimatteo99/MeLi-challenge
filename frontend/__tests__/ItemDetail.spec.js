import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import ItemDetail from "@/components/ItemDetail/ItemDetail";
import { itemDetail } from "./mocks/itemDetailMock";
import { resolvedComponent } from "@/utils/resolvedComponent";

describe("ItemDetail", () => {
  it("should render the component correctly", async () => {
    const item = itemDetail.item;
    const description = itemDetail.description;
    const ItemDetailResolved = await resolvedComponent(ItemDetail, {
      item,
      description
    });
    const { container } = render(<ItemDetailResolved/>);
    expect( container ).toMatchSnapshot();
  });
});
