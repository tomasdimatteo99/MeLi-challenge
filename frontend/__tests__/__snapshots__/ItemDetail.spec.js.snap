// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetail should render the component correctly 1`] = `
<div>
  <div
    class="divContainer"
  >
    <div
      class="categoryContainer"
    >
      <span>
        <span>
          Celulares y Teléfonos
           
           &gt; 
        </span>
      </span>
      <span>
        <span
          class="lastCategorySpan"
        >
          Celulares y Smartphones
        </span>
      </span>
    </div>
    <section
      class="itemDetailContainer"
    >
      <div
        class="leftColumn"
      >
        <div
          class="itemDetailImgContainer"
        >
          <img
            alt="Apple iPhone SE Se (3ª Generación, 64 Gb) - Azul Medianoche - Distribuidor Autorizado"
            class="ImageStyle"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=3840&q=75"
            srcset="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=640&q=75 640w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=750&q=75 750w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=828&q=75 828w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=1080&q=75 1080w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=1200&q=75 1200w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=1920&q=75 1920w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=2048&q=75 2048w, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_696564-MLA52130732644_102022-O.jpg&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
          />
        </div>
        <div
          class="descriptionStyle"
        >
          <div
            class="descriptionTitleContainer"
          >
            <span>
              Descripción del producto
            </span>
          </div>
          <div
            class="descriptionContainer"
          >
            <p>
              Chip A15 Bionic superrápido. Una increíble duración de batería y una cámara que es una superestrella. Y además, el vidrio más resistente en un smartphone y botón de inicio con la seguridad de Touch ID.

Aviso legal

(1) El tamaño de la pantalla se mide en diagonal.
(2) La duración de la batería varía según el uso y la configuración.
(3) El iPhone SE (tercera generación) tiene una clasificación IP67 según la norma IEC 60529 (hasta 30 minutos a una profundidad máxima de 1 metro). La resistencia a las salpicaduras, al agua y al polvo no es una condición permanente, y podría disminuir como consecuencia del uso normal. No intentes cargar un iPhone mojado; consulta el manual del usuario para ver las instrucciones de limpieza y secado. La garantía no cubre daños producidos por líquidos.
(4) Algunas funcionalidades podrían no estar disponibles en todos los países o áreas.
            </p>
          </div>
        </div>
      </div>
      <div
        class="rightColumn"
      >
        <div
          class="dataContainer"
        >
          <div
            class="conditionContainer"
          >
            <h3>
              Nuevo
            </h3>
          </div>
          <div
            class="titleContainer"
          >
            <h1>
              Apple iPhone SE Se (3ª Generación, 64 Gb) - Azul Medianoche - Distribuidor Autorizado
            </h1>
          </div>
          <div
            class="amountContainer"
          >
            <h2>
              $ 633.598,80
            </h2>
          </div>
        </div>
        <div
          class="buttonStyle"
        >
          <button>
            Comprar
          </button>
        </div>
      </div>
    </section>
  </div>
</div>
`;
